### PR TITLE
feat: add websockets feedbacks

### LINKED ISSUE #68: Add backend websocket tests
Add tests for websocket behaviours registered by the core plugin: add end-to-end tests with WS client that connect to the endpoint and subscribe to the channels:

- src/services/items/ws
- src/services/item-memberships/ws

Using branch `68/ws-tests`:

### Issues

- even with mocks, can't seem to connect to WS endpoint (hit `connect ECONNREFUSED 127.0.0.1:3000`): see https://github.com/graasp/graasp/commit/87efde43ae43393eb7e43f8e05b016c8c03adc1e

### LINKED ISSUE #265: changing order does not update via websockets
When you change the order of items, the update is not propagated via websockets.

### LINKED ISSUE #447: [recycle-bin] Sync with recycled items
Original issue: https://github.com/graasp/graasp-plugin-websockets/issues/21

Is it still necessary @dialexo ?

### LINKED ISSUE #547: Add missing `update` hook registration in item service
https://github.com/graasp/graasp/blob/main/src/services/item/service.ts misses some hooks, including `update` (even though the hook is used e.g. at https://github.com/graasp/graasp/blob/45c6b95095f9c219ea2952a449447a3876565c56/src/services/item/ws/hooks.ts#L37

### LINKED ISSUE #579: Recycling and restoring a folder with children does not restore the children
The bug is in the recycle bin service at `restoreMany` since the descendants are marked as soft deleted in the recycle operation (`recycleMany`), but their recovery is omitted in `restoreMany`.

To be fixed in #552

### PR DESCRIPTION
**Pull request by dialexo**
Adds websockets feedback for missing operations

- [x] Closes #447 (and duplicate #266)
- [x] Closes #265
- [x] Closes #579 
- [x] Item updates (on patch), closes #547 
- [x] Closes #68 
- [x] Async operations (e.g. many) + errors
- [x] Add front-end changes

### CODE REVIEW

File: src/services/item/plugins/recycled/ws/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  recycleBinService.hooks.setPostHook('restore', async (member, repositories, { item }) => {
+    const parentId = getParentFromPath(item.path);
+    if (parentId !== undefined) {
+      websockets.publish(itemTopic, parentId, ChildItemEvent('create', item));
+    } else if (item.creator?.id) {
+      // root item, notify creator
+      websockets.publish(memberItemsTopic, item.creator.id, OwnItemsEvent('create', item));
+
+      // TODO: should send to item.creator.id? members that have memberhsip? admins only?
+      websockets.publish(memberItemsTopic, item.creator.id, RecycleBinEvent('delete', item));
```
Comment (swouf) [inline src/services/item/plugins/recycled/ws/hooks.ts:66]: I would say it makes sense to send it to all members who have a membership.

File: src/services/item/plugins/recycled/ws/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    const parentId = getParentFromPath(item.path);
+    if (parentId !== undefined) {
+      websockets.publish(itemTopic, parentId, ChildItemEvent('delete', item));
+    } else if (item.creator?.id) {
+      // root item, notify creator
+      websockets.publish(memberItemsTopic, item.creator.id, OwnItemsEvent('delete', item));
+
+      // TODO: should send to item.creator.id? members that have membership? admins only?
+      websockets.publish(memberItemsTopic, item.creator.id, RecycleBinEvent('create', item));
```
Comment (swouf) [inline src/services/item/plugins/recycled/ws/hooks.ts:36]: I would say it makes sense to send it to all members who have a membership.

**Review by swouf — APPROVED** [review]
Looks good to me. 👍🏼 I think that it would make sense to notify all the users who have a membership with the item. This way, multiple users working on the same folder are really seeing the changes in real-time. What do you think?

File: src/app.ts [hunk 3]
```diff
@@ -75,6 +63,18 @@ export default async function (instance: FastifyInstance): Promise<void> {
   instance.register(async (instance) => {
     // core API modules
     await instance
+      .register(fp(websocketsPlugin), {
```
Comment (pyphilia) [inline src/app.ts:66]: Let's add a comment about this strange bug you had?

**Review by pyphilia — APPROVED** [review]

**Pull request merged**